### ISSUE TITLE
Vidoas: doas.conf is unreadable (macOS)

### ISSUE THREAD

**Issue by FraserEmbrey** [slicer69/doas#57]
I had doas installed previously and it was working fine. I used to be able to run vidoas and save with my password.

Recently I set up my computer with a clean install and now vidoas complains the file is unreadable. The permissions of does.conf are root:admin 600 which I believe to be correct.

If the owner is not root then the file will not work with does `doas: /usr/local/etc/doas.conf is not owned by root`

If the permissions are changed (e.g. root:staff 640) then the file can be opened but on save this will show: `override rw-r-----  root/staff for /usr/local/etc/doas.conf?` If no, the file will not update and if y the file owner will be changed to the current user, breaking doas.

If the permissions are changed to 660 then the file will be writable and the warning will not appear but the file will still save with a non-root owner.

---
**TL;DR**
Can no longer make changes to doas.conf with vidoas. The file will not open without changing the permissions to allow group read access. The file will then save with the user as owner rather than root.

**Comment by slicer69** [comment]
A couple of things here, bear with me, I'm thinking out loud.

1. If you plan to use the vidoas tool then your doas.conf file should be readable to the user you want to be able to edit doas.conf from. It doesn't matter if you do this by adding the user to a specific group or make the doas.conf file world-readable. Either way the user needs to be able to read the file. Otherwise they can't see/edit it.

2. I cannot duplicate the issue you describe with the doas.conf file being saved as owned by another user. When I run vidoas the resulting doas.conf file is always owned by root. It's working on Linux and FreeBSD as expected, which makes me think this must be a macOS-specific issue, which I cannot trouble-shoot.

3. I'm guessing a macOS specific patch could be introduced to the vidoas script which changes the ownership of the doas.conf file. I'm going to try making an adjustment to the vidoas script and see if we can make it more macOS friendly, but I won't be able to test it. Perhaps you can try it out and report back after I make the change?

**Comment by slicer69** [comment]
I have made some updates to the vidoas script. There are a few key changes:

1. The script should now work, regardless of the permissions on the doas.conf file.
2. The vidoas script must be run with root access, probably from doas itself. ie doas vidoas
3. The ownership of the doas.conf file should not be preserved.

Please let me know how it works out.

**Comment by FraserEmbrey** [comment]
Thank you very much. I have tried with the latest version.

doas.conf was set as root:staff 640. Without doas/sudo vidoas opened the file but gave this error on save: `install: /usr/local/etc/doas.conf: chown/chgrp: Operation not permitted` the doas.conf file was then deleted.

With root it does work fine, although it was working with root permission before anyway since it would save as the root user.

Trying to run vidoas a second time gives this message: `doas: doas is not enabled, /usr/local/etc/doas.conf: No such file or directory`.

It definitely used to work before without root. I don't think the issue is macOS 11 related as I am sure I was using doas with macOS 11 during the beta period.

I remember before being able to type vidoas as a non-root user and using my password to save the file.

**Comment by slicer69** [comment]
An earlier version of the script did perform some actions, like editing and performing initial checks, without root access. It would prompt for a password if needed. However, we've changed that. The current version requires root as, by default, we now make the doas.conf file accessible only by root. (Ownership root, permissions 0600.)

Basically now the vidoas script needs to run as root or with a tool like sudo/doas.

I'm a bit concerned about the part you mentioned about doas.conf being deleted. There shouldn't be anything in the vidoas script which could remove the original file. And, if vidoas was run without sudo/doas/root access then that means your user was able to remove the configuration file even though they did not have permission to write to it (permission 640 set on the file). Basically any tool shouldn't be able to remove a file you don't have access to so something and we don't try to remove the original file so multiple things must have gone really wrong there.

**Comment by FraserEmbrey** [comment]
That concerns me too. I don't know how that can happen and assume the doas script can access root permissions somehow since this happens repeatedly.

If I run `doas vidoas` then all works fine.

**Comment by FraserEmbrey** [comment]
Here is a screenshot from my terminal, I used a root shell to recreate a doas.conf file from the backup and alter the permissions:

```
# cp /usr/local/etc/doas.conf{.bkp,}
# chmod 644 /usr/local/etc/doas.conf
```
<img width="663" alt="Screenshot" src="https://user-images.githubusercontent.com/7635426/105881791-d4b59780-5ffc-11eb-8c6b-e996b8c2cc02.png">

**Comment by slicer69** [comment]
The vidoas script does not have doas access built into it, it can't do anything beyond what your regular user can do. It doesn't run doas or have setuid powers or anything like that. Which means your regular user can delete (and probably modify) your doas.conf file, even with permissions set to 640. This is something that should get checked out as there may be ACL permissions at play granting your user extra access it shouldn't have.

Glad to hear the "doas vidoas" command is working as expected. I've updated the manual page to reflect that root access is needed for vidoas in its current form.

**Comment by slicer69** [comment]
I think I may know what is happening. I'm pretty sure your regular user account has write access to the /usr/local/etc/ directory. That would explain why it can overwrite doas.conf without doas/root access.

When I try to run vidoas on my machine as my regular user it immediately errors out saying 

     vidoas: /usr/local/etc is not writable
     vidoas: You probably need to run vidoas as root

If you are not seeing that then your user probably has write access to /usr/local/etc and that explains why it can erase a file it doesn't have write access for.

**Comment by FraserEmbrey** [comment]
I did consider if there were ACL permissions but there don't seem to be any. I am very confused with
1. why the file is deleted by vidoas
2. why I seem to have permission to make any changes to files there

Surely my user shouldn't be able to delete files if they're not writable, perhaps I should remove my access from the directory.

**Comment by slicer69** [comment]
On UNIX systems having write access to the directory will allow your user to remove files inside the directory. However, it is really unusual for non-root users to have write access to any part of the /usr structure, including /usr/local/etc.

**Comment by FraserEmbrey** [comment]
I have read/write access as I use homebrew which sets itself up in /usr/local/

I don't particularly need access to the etc folder as far as I know though.

**Comment by slicer69** [comment]
A regular user having write access is definitely a problem. It means users can delete and replace the doas.conf file as they like which basically grants everyone root access if they want it.

You should probably also make /usr/local/bin read-only as well, otherwise anyone on your system can remove and replace the doas command (or any other command) with whatever they want. That makes it really easy for someone to harvest your password.

**Comment by slicer69** [comment]
Since the original issue is resolved I'm going to make this ticket as closed. Feel free to open a new one if the adjusted permissions on /usr/local don't fix the remaining problem with doas.conf editing.

**Comment by FraserEmbrey** [comment]
Thank you for your help
